### PR TITLE
Resilience diagnostics

### PR DESCRIPTION
**Pull request by slavapestov**
rdar://problem/29164469
rdar://problem/23522157

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test macOS

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test Linux

File: lib/Sema/ResilienceDiagnostics.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+enum FragileFunctionKind : unsigned {
+  Transparent,
+  InlineAlways,
+  Inlineable,
+  DefaultArgument
+};
+
+/// If the body is able to be inlined into functions in other resilience
+/// domains, this ensures that only sufficiently-conservative access patterns
```
Comment (jrose-apple) [inline lib/Sema/ResilienceDiagnostics.cpp]: Nitpick: "access patterns" isn't a Sema-level term.

File: test/SILOptimizer/devirt_access.swift [hunk 2]
```diff
@@ -4,18 +4,6 @@
 // RUN: %target-swift-frontend -O -primary-file %s %S/Inputs/devirt_access_helper.swift -I %t -emit-sil -sil-inline-threshold 1000 -sil-verify-all | %FileCheck -check-prefix=WHOLE-MODULE %s
 // RUN: %target-swift-frontend -O -primary-file %s %S/Inputs/devirt_access_helper.swift -I %t -emit-sil -sil-inline-threshold 1000 -sil-verify-all | %FileCheck -check-prefix=PRIMARY-FILE %s
 
-import devirt_access_other_module
```
Comment (jrose-apple) [inline test/SILOptimizer/devirt_access.swift]: I don't think it's a good idea to just delete part of the test. Maybe run it under explicit non-resilient mode?

File: include/swift/AST/DiagnosticsSema.def [hunk 3]
```diff
@@ -3296,6 +3296,35 @@ NOTE(availability_conformance_introduced_here, none,
...
+  "an '@inline(__always)' function|" \
+  "an '@_inlineable' function|" \
+  "a default argument value}"
+
+ERROR(local_type_in_inlineable_function,
+      none, "type %0 cannot be nested inside " FRAGILE_FUNC_KIND "1",
+      (DeclName, unsigned))
+
+ERROR(resilience_decl_unavailable,
+      none, "%0 %1 is %select{internal|private}2 and cannot be referenced "
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: This leaves out `fileprivate`. Can you use Accessibility as the argument here, even if `public` and `open` don't make sense?

File: include/swift/AST/DiagnosticsSema.def [hunk 4]
```diff
@@ -3296,6 +3296,35 @@ NOTE(availability_conformance_introduced_here, none,
      "conformance introduced here", ())
 
 //------------------------------------------------------------------------------
+// Resilience diagnostics
+//------------------------------------------------------------------------------
+
+ERROR(versioned_attr_with_explicit_accessibility,
+      none, "'@_versioned' attribute can only be applied to internal "
+      "declarations, but %0 is %select{private|fileprivate|internal|public}1",
+      (Identifier, unsigned))
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: This should use Accessibility as the argument kind, and I suggest also capitalizing "INTERNAL" and "PUBLIC" (and "OPEN"?) to show that they should never occur.

**Review by jrose-apple — COMMENTED** [review]
Glad to see diagnostics coming in! I'm a little uncomfortable to see this being defined in terms of ResilienceExpansion, though. I maintain that we're going to need AvailabilityContext as soon as we have versioned attributes, and I'd rather see ResilienceExpansion squashed and rebuilt from scratched than extended to match our current resilience model. (…because I don't feel like there is a natural way to do such an extension.)

File: test/SILOptimizer/devirt_access.swift (hunk 2 above)
Comment (slavapestov) [inline test/SILOptimizer/devirt_access.swift]: The "explicit non-resilient" mode would just disable the new diagnostics. I'd be OK with making them conditional if we really had to, but I think basically this test was just exercising that the SIL optimizer doesn't blow up in a case that no longer type checks.

File: include/swift/AST/DiagnosticsSema.def (hunk 4 above)
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def]: Thanks, done!

File: include/swift/AST/DiagnosticsSema.def (hunk 3 above)
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def]: Ditto

File: lib/Sema/ResilienceDiagnostics.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/Sema/ResilienceDiagnostics.cpp]: I just removed the comment since it got copy-pasted from getResilienceExpansion() on accident.

**Comment by slavapestov** [comment]
@jrose-apple I plan on re-visiting the whole ResilienceExpansion thing soon. I wanted to prototype the semantics and get some tests in first, though. I made the other fixes you suggested -- are you OK with this getting merged?

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Yeah, I guess it's more important to start diagnosing things right now. I'll admit part of my discomfort is that ResilienceExpansion doesn't naturally make sense to me, even though it's such a simple enum. But *that* part shouldn't block you.

**Pull request merged**